util_for_test.go

```diff
@@ -0,0 +1,329 @@
+			if err != nil {
+				t.Fatalf("failed to select: %v", err)
+			}
+			jsonBytes, err := json.Marshal(data)
+			if err != nil {
+				t.Fatalf("failed to marshal to json: %v", err)
+			}
+			// Note the Index field, which would change based on the for loop.
+			type dataholder struct {
+				ID    string `json:"id,omitempty"`
+				Name  string `json:"name,omitempty"`
+				Index int    `json:"index,omitempty"`
+			}
+			d := []dataholder{}
+			err = json.Unmarshal(jsonBytes, &d)
+			if err != nil {
+				t.Fatalf("failed to unmarshal data: %v", err)
+			}
+			// Simply check all fields match as expected.
+			want := dataholder{ID: "user:x", Name: "x", Index: i}
+			if d[0].ID != want.ID || d[0].Name != want.Name || d[0].Index != want.Index {
+				t.Fatalf("data mismatch:\n    want: %+v, got:  %+v", want, d)
+			}
+		})
+	}
+}
+
+type DBForTest struct {
+	*DB
+}
+
+func NewTestSurrealDB(t testing.TB) (string, *DBForTest, func()) {
+	t.Helper()
+
+	checkSurrealCLIOnce.Do(func() {
+		if _, err := exec.LookPath("surreal"); err == nil {
+			testHasSurrealCLI = true
+		}
+	})
+	if !testHasSurrealCLI {
+		t.Skip("surreal CLI not found, skipping")
+	}
+
+	port := getFreePort(t)
+	srvEndpoint := fmt.Sprintf("0.0.0.0:%d", port)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cmd := exec.CommandContext(ctx, "surreal",
+		"start",
+		"--auth",
+		"--user", "root",
+		"--pass", "surrealdb",
+		"--bind", srvEndpoint,
+		"memory")
+	// Ensure to wait for a short while before context cancellation to
+	// propagate, so that the client can shut down before.
+	// NOTE: There is currently a race condition and potential panic due to the
+	// closed websocket being used in a tight loop.
+	cmd.WaitDelay = DelayBeforeServerExit
+
+	go func() {
+		err := cmd.Run()
+		if err != nil {
+			if errors.Is(ctx.Err(), context.Canceled) {
+				t.Logf("SurrealDB server shut down by context")
+				return
+			}
+			exit := &exec.ExitError{}
+			if errors.As(err, &exit) {
+				t.Errorf("SurrealDB server exited with error: %v", err)
+				return
+			}
+			t.Errorf("Command failed to run: %v", err)
+		}
+	}()
+
+	clientEndpoint := fmt.Sprintf("localhost:%d", port)
+
+	// Give some time for the server to start up.
+	time.Sleep(DelayAfterServerStart)
+	db := newDBForTest(t, clientEndpoint)
+	login := &Auth{
+		Username:  "root",
+		Password:  "surrealdb",
+		Namespace: "dummy",
+		Database:  "dummy",
+	}
+	if _, err := db.Signin(login); err != nil {
+		t.Fatalf("failed to connect using '%+v' for signin: %v", login, err)
+	}
+
+	// Client can potentially stay around, so ensure to close connection after
+	// the test is complete.
+	t.Cleanup(func() {
+		db.conn.Close()
+	})
+
+	return clientEndpoint, &DBForTest{db}, cancel
+}
+
+func getFreePort(t testing.TB) uint32 {
+	t.Helper()
+
+	addr, err := net.ResolveTCPAddr("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("could not find any open port: %v", err)
+	}
+	l, err := net.ListenTCP("tcp", addr)
```

Comment:
That's what I do just a few lines above -- but the complication was coming from how I was using `net.ListenTCP` instead of `net.Listen`. I have simplified this code to use localhost:0 directly 6afa007